build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1188,6 +1212,34 @@ message UpdateActionResultRequest {
   ResultsCachePolicy results_cache_policy = 4;
 }
 
+// A request message for
+// [RepositoryCache.Download][build.bazel.remote.execution.v2.RepositoryCache.Download].
+message DownloadRequest {
+  // The instance of the execution system to operate against. A server may
+  // support multiple instances of the execution system (with their own workers,
+  // storage, caches, etc.). The server MAY require use of this field to select
+  // between them in an implementation-defined fashion, otherwise it can be
+  // omitted.
+  string instance_name = 1;
+
+  // List of mirror URLs referencing the same file.
+  repeated string urls = 2;
+
+  // Expected checksum of the file downloaded, in Subresource Integrity format.
```

Comment:
There's an interesting tension here between using SRI (which is nicely standard, and which e.g. maven .sha1 files can be used to populate even though they don't have known size, and the digest function may be different), vs using a `Digest` (which fits in nicely with the rest of the remote execution ecosystem)... If we really want to have SRI here, maybe this could become a `OneOf` field so that if you _do_ know the digest, that information can be used as an optimisation?

Reply:
We can't use a `Digest` for the input because that would require knowing the size, which (at least in Bazel) is not something users are expected to provide for all external files.

Reply:
I think what Daniel's suggesting here makes sense, and in fact generalizes further - broadly speaking a DownloadRequest may ultimately be used for all sorts of resources, with all sorts of integrity schemes. SRI makes sense for some purposes. Digest makes sense for others (including, I'll note, as a way to request server-side cross-CAS asset population). But for a generic API we should assume that others will ultimately be desired as well, and prepare to gracefully handle that.

I see three main options for addressing that:

 1.  Make the 'integrity' field optional, and allow other options as they come as their own optional top-level fields.
 2. Use a OneOf to express the same, but be clear that they're mutually exclusive.
 3. Use an Any to express the same.

I don't really like 1) - multiple fields - because it requires a proliferation of fields, and if anyone tries to use it privately and then open up later we risk field collisions.
I don't really like 2) - oneof type - either, because integrity formats are not actually exclusive; a client that knows multiple should be able to specify both, e.g. to ease graceful transitions.
Which leads me to 3) - Any type - as my preferred option. I.e. something like:

```
message DownloadRequest {
  ...
  // Optional integrity message, for doing server-side validation that the resource to be download was
  // correctly fetched.
  // Servers MAY support any formats they wish, or none. Common formats include `Digest` and 
  // `SubresourceIntegrity`.
  // If any messages are specified, servers MUST validate at least one matched or return an error.
  // <what error specifically?>
  repeated google.protobuf.Any integrity = 3;
}

message SubresourceIntegrity {
  string integrity = 1;
}
```

Reply:
If a future PR adds support for features like cache prefill, I'd rather have that in its own RPC rather than try to multiplex a half-dozen behaviors into this one. For example, cross-CAS asset distribution would use an RPC in `ContentAddressableStorage`.

Regarding specifically `Digest` vs SRI: the purpose of the repository cache is to handle ingestion of files that are not yet present in the build cluster and, therefore, do not have a `Digest`. If the client already has a `Digest` then it does not need to call `RepositoryCache.Download()`, it can simply pass that `Digest` to `ContentAddressableStorage`.

Reply:
In Pants, we're looking to use `Digest` as our remote file unit of truth, and that may involve things not already in the `ContentAddressableStorage`. Our primitive here would be something like:
```
remote_file(
  digest = Digest(fingerprint=..., length=...),
  urls = [...],
)
```

but we want to be able to pass this to the remote execution system to fetch so that if the client is a mac and the execution platform is linux, we don't need to round-trip a linux-only clang via the mac which can't possibly use it, just to upload it to the remote execution server.

Reply:
@illicitonion If you have a `Digest` then it can be converted to an SRI hash without much trouble. Bazel doesn't require users to specify sizes of remote files, so if this API was specified in terms of `Digest` it would save you a very small amount of code at the cost of not working at all for my Bazel use case.

Reply:
@jmillikin-stripe maybe I should take a step back. In the narrow form of _"only fetch single files"_ and _"only address content by URL+SRI"_, I don't think I'd support having this interface added here. The concerns I have boil down to:

   * In this form it'd have the effect of entrenching what I consider to be an antipattern. For bazel to take a dependency on a git repo right now it must:
      1. Be given the hash of a tarball archive of that repo. _(Cost: user experience...they're being forced to reference something derived from what they wanted to express, which was repo+hash. Also, of tangential interest: https://github.com/easybuilders/easybuild-easyconfigs/issues/5151)_
      1. Download that entire archive. _(Cost: network and time. This phase already takes whole minutes for some repos, and scales with the size of the dependencies rather than the size of the built subset.)_
      1. Unpack that entire archive. _(Cost: disk IO. Puts a floor on the minimum resources required locally at the build host.)_
      1. Load the BUILD and .bzl files contained in that archive.
      1. Re-hash everything, to calculate its digest. (Cost: disk IO and CPU).

     That's a lot of unscalable work, most of it unnecessary, and causing real-world performance bottlenecks today. I'm happy to take incremental steps towards improving it - repo fetching via a local cache would reduce the time needed for the download step, for example - but only if they enable us to keep iterating further. A single-purpose API as you describe would actually get in the way of that iteration: it wouldn't make it easier to change either end of (how dependencies are expressed; how they're returned), and using something else would then require clients to support _multiple_ APIs and pick between them.
   * In this context - this repo, and set of APIs - it feels like an awkward fit. Neither URLs nor SRI are native concepts for a gRPC API, or this ecosystem in particular. They make sense as a compatibility shim specifically for downloading files from websites, but that in turn makes it less suited for any other purpose. And this repo is all about having a small set of flexible, powerful APIs that cross clients and servers and use-cases.

But that all said, I _do_ believe we can add it in a form that both supports your purposes - indirecting downloads of repository cache files in bazel - and mine - iterating towards APIs clients can generally use to download and access remote files. For you it does come with the cost of added indirection and generality, but I'm hoping that's acceptable enough.

Reply:
@EricBurnett OK, do you think I should move this new service spec to bazelbuild/bazel then?

This spec is not designed to solve issues with slow commands executed during repository setup, so even if I changed it to fit your goals, I would have to submit a separate proposal to get this same download-oriented API in for my purposes.

Note that for me, blocking network access from arbitrary command execution is an explicit goal and I do not want to support patterns like `git_repository()` here.

Reply:
I'm not sure I understand your concern - making the API more general does not require you to actually *implement* support for anything else, and certainly not git repos or multi-file returns. I imagined you'd still only implement support for accepting SRI and returning the file's Digest, as you desire, and in the first cut that's all bazel would implement either. I did not intend to suggest you change what features you implement (now or later), beyond how they're expressed in the on-the-wire protos.

But that all said, if you prefer this as a narrow bazel API to solve your specific pain point, I think bazelbuild/bazel is a fine home for it. Someone on the bazel side will have to ok it though, I cannot speak for them.